src/detect-base64-decode.c

```diff
@@ -0,0 +1,723 @@
+/* Copyright (C) 2015 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#undef DEBUG_BASE64
+
+#include "suricata-common.h"
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-base64-decode.h"
+#include "util-base64.h"
+#include "util-print.h"
+
+/* Arbitrary maximum buffer size for decoded base64 data. */
+#define BASE64_DECODE_MAX 65535
+
+static const char decode_pattern[] = "\\s*(bytes\\s+(\\d+),?)?"
+    "\\s*(offset\\s+(\\d+),?)?"
+    "\\s*(\\w+)?";
+static pcre *decode_re = NULL;
+
+static int DetectBase64DecodeSetup(DetectEngineCtx *, Signature *, char *);
+static void DetectBase64DecodeFree(void *);
+static void DetectBase64DecodeRegisterTests(void);
+
+void DetectBase64DecodeRegister(void)
+{
+    const char *pcre_errptr;
+    int pcre_erroffset;
+
+    sigmatch_table[DETECT_BASE64_DECODE].name = "base64_decode";
+    sigmatch_table[DETECT_BASE64_DECODE].desc =
+        "Decodes base64 encoded data.";
+    sigmatch_table[DETECT_BASE64_DECODE].url =
+        "https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Payload_keywords#base64_decode";
+    sigmatch_table[DETECT_BASE64_DECODE].Setup = DetectBase64DecodeSetup;
+    sigmatch_table[DETECT_BASE64_DECODE].Free = DetectBase64DecodeFree;
+    sigmatch_table[DETECT_BASE64_DECODE].RegisterTests =
+        DetectBase64DecodeRegisterTests;
+
+    sigmatch_table[DETECT_BASE64_DECODE].flags |= SIGMATCH_PAYLOAD;
+    sigmatch_table[DETECT_BASE64_DECODE].flags |= SIGMATCH_OPTIONAL_OPT;
+
+    decode_re = pcre_compile(decode_pattern, 0, &pcre_errptr, &pcre_erroffset,
+        NULL);
+    if (decode_re == NULL) {
+        SCLogError(SC_ERR_PCRE_COMPILE, "Failed to compile pattern \"%s\" at"
+            " offset %d: %s", decode_pattern, pcre_erroffset, pcre_errptr);
+        exit(EXIT_FAILURE);
+    }
```

Comment:
It couldn't optimize the query - it returns NULL, with an err or NULL.  I've added it back for consistency.